README.MD
This is an online community auto-reply chatbot framework. It includes text classification module, text generation module, and deployment script. Users can quickly build their own community auto-reply chatbot. They only need to download this repo, configure the environment on their own machine,  import their own data sets, and fill in their community API.
##### The workflow of the framework
1. Fetch the latest posts through community API
2. Classify the type of the posts(this type can be used as the basis for judging whether to reply in the next step)
3. Generate comments
4. Reply automatically in the community through community API

**Reference**

OpenNMT：https://github.com/OpenNMT/OpenNMT-py

CNN Classfier：https://github.com/gaussic/text-classification-cnn-rnn

# Step1:Setup

Requirements：

- Python >= 3.5
- Torch == 1.0.0
- Torchvision == 0.2.1
- Torchtext == 0.4.0

Install ``` onmt ``` from ```  OpenNMT/setup.py ```:
``` 
python setup.py install
 ```
# Step2:Prepare Your Dataset

##### 1.For text classification model

Prepare four files in ``` Classifier/data/cnews/ ``` ：
- data.train.txt
- data.val.txt
- data.test.txt
- data.pred.txt

##### 2.For text generation model

Prepare following files in ``` OpenNMT/data/ ``` ：
- src-train.txt
- src-val.txt
- src-test.txt
- tgt-train.txt
- tgt-val.txt

# Step3:Train the Classification Model

##### 1.CNN parameter in Classifier/cnn_model.py
 ```
class TCNNConfig(object):

    embedding_dim = 64  
    seq_length = 600  
    num_classes = 2  
    num_filters = 256  
    kernel_size = 5  
    vocab_size = 5000  

    hidden_dim = 128  

    dropout_keep_prob = 0.5  
    learning_rate = 1e-3  

    batch_size = 64  
    num_epochs = 100  

    print_per_batch = 10  
    save_per_batch = 10 
 ```

##### 2.  Train the model

In ``` Classifier/``` directory,  run ``` python run_cnn.py train ``` , now it start training

After running the training, the following files are generated in ```Classifier/data/cnews/```:

- `data.vocab.txt`

##### 3. Test the model

In ``` Classifier/``` directory, run ```python run_cnn.py test``` to test on ```data.test.txt```

##### 4. Predict

```Classifier/predict.py``` provide predict function of CNN model. Run ```predict.py``` to predict sentence on ```Classifer/data/cnews/data.predict.txt```. This will output predictions into ```Classifier/predict.txt```.

# Step4:Train the Generation Model

##### 1. Preprocess the data

run ```OpenNMT/preprocess.py```

```
python preprocess.py -train_src data/src-train.txt -train_tgt data/tgt-train.txt -valid_src data/src-val.txt -valid_tgt data/tgt-val.txt -save_data data/demo
```

Validation files are required and used to evaluate the convergence of the training. It usually contains no more than 5000 sentences.


After running the preprocessing, the following files are generated in ```OpenNMT/data/```:

* `demo.train.pt`: serialized PyTorch file containing training data
* `demo.valid.pt`: serialized PyTorch file containing validation data
* `demo.vocab.pt`: serialized PyTorch file containing vocabulary data


Internally the system never touches the words themselves but uses these indices.

##### 2. Train the model

run ```OpenNMT/train.py```

```bash
python train.py -data data/demo -save_model demo-model
```

The main train command is quite simple. Minimally it takes a data file and a save file.  This will run the default model, which consists of a 2-layer LSTM with 500 hidden units on both the encoder/decoder. If you want to train on GPU, you need to set, as an example: CUDA_VISIBLE_DEVICES=1,3 `-world_size 2 -gpu_ranks 0 1` to use (say) GPU 1 and 3 on this node only. To know more about distributed training on single or multi nodes, read the FAQ section:xxxxxxx

##### 3. Translate

run ```OpenNMT/translate_original.py```

```bash
python translate_original.py -model demo-model_acc_XX.XX_ppl_XXX.XX_eX.pt -src data/src-test.txt -output pred.txt -replace_unk -verbose
```
Now you have a model that you can use to predict on new data. We do this by running beam search. This will output predictions into `pred.txt`.

# Step5: Run Deployment Script

##### 1.Set  API and parameter

In ```OpenNMT/Deployment.py``` file, you can fill in your own Url, parameter, and simulative user information:
```
#################################################################################################
##############You should fill in your community API and simulative user information##############
#########################and modify time parameter if you want###################################

THRESHOLD = 10  # the threshold for deciding whether the chatbot needs to respond to the overlooked post or not 
STUDY_TIME = 60 * 24 * 7  # the whole deployment period 
OBSERVE_INTERVAL = 9  # the interval time between getting latest posts 
COMMENT_INTERVAL = 2  # the interval time detecting if observed posts have been replied 

Community_getLatestPost_Url = ""
Community_toComment_Url = ""
Community_getPostDetail_Url = ""


AI_auth_list = [["username1", "<authorization1>"],
                ["username2", "<authorization2>"],
                ["username3", "<authorization3>"]]
# e.g.
# username = "saltone"
# authorization = "XDS 7.fIC1Fkcg6-Qa6--o9qUP-FyrhLkyLLZOMN6r7Jxxx"

#################################################################################################
##############You should fill in your community API and simulative user information##############
#########################and modify time parameter if you want###################################
```
##### 2. Run deployment script

run ``` OpenNMT/Deployment.py ```

In console , you will see following log if you did not do anything :
```
content: This is a post example
comment: This is a comment example
Do you agree to Comment? Input nothing to confirm or input an appropriate sentence:
 Agree to comment
chatbot will comment on this sentence: This is a comment example
```
If you input a new sentence, the comment will be refined:
```
content: This is a post example
comment: This is a comment example
Do you agree to Comment? Input nothing to confirm or input an appropriate sentence: Fighting!!!
chatbot will comment on this sentence: Fighting!!!
```

# Note

1.Different online communities have different APIs and require different parameters. This part needs to be modified according to the specific situation.

2.OpenNMT has been updated to version 1.7, which is not compatible with the version（1.0.0） used in this repo.

3.If you have any questions, please contact me by email:wangliuping17@mails.ucas.ac.cn


OpenNMT/CHANGELOG.md

**Notes on versioning**


## [Unreleased]
### Fixes and improvements

## [0.8.2](https://github.com/py/tree/0.8.2) (2019-02-16)
* Update documentation and Library example
* Revamp args
* Bug fixes, save moving average in FP32
* Allow FP32 inference for FP16 models

## [0.8.1](https://github.com/py/tree/0.8.1) (2019-02-12)
* Update documentation
* Random sampling scores fixes
* Bug fixes

## [0.8.0](https://github.com/py/tree/0.8.0) (2019-02-09)
* Many fixes and code cleaning thanks @flauted, @guillaumekln
* Datasets code refactor (thanks @flauted) you need to r-preeprocess datasets

### New features
* FP16 Support: Experimental, using Apex, Checkpoints may break in future version.
* Continuous exponential moving average (thanks @francoishernandez, and Marian)
* Relative positions encoding (thanks @francoishernanndez, and Google T2T)
* Deprecate the old beam search, fast batched beam search supports all options


## [0.7.2](https://github.com/py/tree/0.7.2) (2019-01-31)
* Many fixes and code cleaning thanks @bpopeters, @flauted, @guillaumekln

### New features
* Multilevel fields for better handling of text featuer embeddinggs. 


## [0.7.1](https://github.com/py/tree/0.7.1) (2019-01-24)
* Many fixes and code refactoring thanks @bpopeters, @flauted, @guillaumekln

### New features
* Random sampling thanks @daphnei
* Enable sharding for huge files at translation

## [0.7.0](https://github.com/py/tree/0.7.0) (2019-01-02)
* Many fixes and code refactoring thanks @benopeters
* Migrated to Pytorch 1.0

## [0.6.0](https://github.com/py/tree/0.6.0) (2018-11-28)
* Many fixes and code improvements
* New: Ability to load a yml config file. See examples in config folder.

## [0.5.0](https://github.com/py/tree/0.5.0) (2018-10-24)
* Fixed advance n_best beam in translate_batch_fast
* Fixed remove valid set vocab from total vocab
* New: Ability to reset optimizer when using train_from
* New: create_vocabulary tool + fix when loading existing vocab.

## [0.4.1](https://github.com/py/tree/0.4.1) (2018-10-11)
* Fixed preprocessing files names, cleaning intermediary files.

## [0.4.0](https://github.com/py/tree/0.4.0) (2018-10-08)
* Fixed Speech2Text training (thanks Yuntian)

* Removed -max_shard_size, replaced by -shard_size = number of examples in a shard.
  Default value = 1M which works fine in most Text dataset cases. (will avoid Ram OOM in most cases)


## [0.3.0](https://github.com/py/tree/0.3.0) (2018-09-27)
* Now requires Pytorch 0.4.1

* Multi-node Multi-GPU with Torch Distributed

  New options are:
  -master_ip: ip address of the master node
  -master_port: port number of th emaster node
  -world_size = total number of processes to be run (total GPUs accross all nodes)
  -gpu_ranks = list of indices of processes accross all nodes

* gpuid is deprecated
See examples in https://github.com/py/blob/master/docs/source/FAQ.md

* Fixes to img2text now working

* New sharding based on number of examples

* Fixes to avoid 0.4.1 deprecated functions.


## [0.2.1](https://github.com/py/tree/0.2.1) (2018-08-31)

### Fixes and improvements

* First compatibility steps with Pytorch 0.4.1 (non breaking)
* Fix TranslationServer (when various request try to load the same model at the same time)
* Fix StopIteration error (python 3.7)

### New features
* Ensemble at inference (thanks @Waino)

## [0.2](https://github.com/py/tree/v0.2) (2018-08-28)

### improvements

* Compatibility fixes with Pytorch 0.4 / Torchtext 0.3
* Multi-GPU based on Torch Distributed
* Average Attention Network (AAN) for the Transformer (thanks @francoishernandez )
* New fast beam search (see -fast in translate.py) (thanks @guillaumekln)
* Sparse attention / sparsemax (thanks to @bpopeters)
* Refactoring of many parts of the code base:
 - change from -epoch to -train_steps -valid_steps (see opts.py)
 - reorg of the logic train => train_multi / train_single => trainer
* Many fixes / improvements in the translationserver (thanks @pltrdy @francoishernandez)
* fix BPTT

## [0.1](https://github.com/py/tree/v0.1) (2018-06-08)

### First and Last Release using Pytorch 0.3.x




OpenNMT/Deployment.py
import datetime
import random
import re
import time
import urllib
import chardet
import ssl
import signal


from pandas.io.json import json

from Classifer.predict import CnnModel
from OpenNMT.translate import interface

ssl._create_default_https_context = ssl._create_unverified_context







class InputTimeoutError(Exception):
    pass

def interrupted(signum, frame):
    raise InputTimeoutError

signal.signal(signal.SIGALRM, interrupted)

signal.alarm(10)#


#################################################################################################
##############You should fill in your community API and simulative user information##############
#########################and modify time parameter if you want###################################

THRESHOLD = 10  # the threshold for deciding whether the chatbot needs to respond to the overlooked post or not
STUDY_TIME = 60 * 24 * 7  # the whole deployment period
OBSERVE_INTERVAL = 9  # the interval time between getting latest posts
COMMENT_INTERVAL = 2  # the interval time detecting if observed posts have been replied


Community_getLatestPost_Url = ""
Community_toComment_Url = ""
Community_getPostDetail_Url = ""


AI_auth_list = [["username1", "<authorization1>"],
                ["username2", "<authorization2>"],
                ["username3", "<authorization3>"]]
# e.g.
# username = "saltone"
# authorization = "XDS 7.fIC1Fkcg6-Qa6--o9qUP-FyrhLkyLLZOMN6r7Jxxx"

#################################################################################################
##############You should fill in your community API and simulative user information##############
#########################and modify time parameter if you want###################################


user_agent_list = [
    "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1; SV1; AcooBrowser; .NET CLR 1.1.4322; .NET CLR 2.0.50727)",
    "Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 6.0; Acoo Browser; SLCC1; .NET CLR 2.0.50727; Media Center PC 5.0; .NET CLR 3.0.04506)",
    "Mozilla/4.0 (compatible; MSIE 7.0; AOL 9.5; AOLBuild 4337.35; Windows NT 5.1; .NET CLR 1.1.4322; .NET CLR 2.0.50727)",
    "Mozilla/5.0 (Windows; U; MSIE 9.0; Windows NT 9.0; en-US)",
    "Mozilla/5.0 (compatible; MSIE 9.0; Windows NT 6.1; Win64; x64; Trident/5.0; .NET CLR 3.5.30729; .NET CLR 3.0.30729; .NET CLR 2.0.50727; Media Center PC 6.0)",
    "Mozilla/5.0 (compatible; MSIE 8.0; Windows NT 6.0; Trident/4.0; WOW64; Trident/4.0; SLCC2; .NET CLR 2.0.50727; .NET CLR 3.5.30729; .NET CLR 3.0.30729; .NET CLR 1.0.3705; .NET CLR 1.1.4322)",
    "Mozilla/4.0 (compatible; MSIE 7.0b; Windows NT 5.2; .NET CLR 1.1.4322; .NET CLR 2.0.50727; InfoPath.2; .NET CLR 3.0.04506.30)",
    "Mozilla/5.0 (Windows; U; Windows NT 5.1; zh-CN) AppleWebKit/523.15 (KHTML, like Gecko, Safari/419.3) Arora/0.3 (Change: 287 c9dfb30)",
    "Mozilla/5.0 (X11; U; Linux; en-US) AppleWebKit/527+ (KHTML, like Gecko, Safari/419.3) Arora/0.6",
    "Mozilla/5.0 (Windows; U; Windows NT 5.1; en-US; rv:1.8.1.2pre) Gecko/20070215 K-Ninja/2.1.1",
    "Mozilla/5.0 (Windows; U; Windows NT 5.1; zh-CN; rv:1.9) Gecko/20080705 Firefox/3.0 Kapiko/3.0",
    "Mozilla/5.0 (X11; Linux i686; U;) Gecko/20070322 Kazehakase/0.4.5",
    "Mozilla/5.0 (X11; U; Linux i686; en-US; rv:1.9.0.8) Gecko Fedora/1.9.0.8-1.fc10 Kazehakase/0.5.6",
    "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/535.11 (KHTML, like Gecko) Chrome/17.0.963.56 Safari/535.11",
    "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_7_3) AppleWebKit/535.20 (KHTML, like Gecko) Chrome/19.0.1036.7 Safari/535.20",
    "Opera/9.80 (Macintosh; Intel Mac OS X 10.6.8; U; fr) Presto/2.9.168 Version/11.52",
    "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/536.11 (KHTML, like Gecko) Chrome/20.0.1132.11 TaoBrowser/2.0 Safari/536.11",
    "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.1 (KHTML, like Gecko) Chrome/21.0.1180.71 Safari/537.1 LBBROWSER",
    "Mozilla/5.0 (compatible; MSIE 9.0; Windows NT 6.1; WOW64; Trident/5.0; SLCC2; .NET CLR 2.0.50727; .NET CLR 3.5.30729; .NET CLR 3.0.30729; Media Center PC 6.0; .NET4.0C; .NET4.0E; LBBROWSER)",
    "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1; SV1; QQDownload 732; .NET4.0C; .NET4.0E; LBBROWSER)",
    "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/535.11 (KHTML, like Gecko) Chrome/17.0.963.84 Safari/535.11 LBBROWSER",
    "Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 6.1; WOW64; Trident/5.0; SLCC2; .NET CLR 2.0.50727; .NET CLR 3.5.30729; .NET CLR 3.0.30729; Media Center PC 6.0; .NET4.0C; .NET4.0E)",
    "Mozilla/5.0 (compatible; MSIE 9.0; Windows NT 6.1; WOW64; Trident/5.0; SLCC2; .NET CLR 2.0.50727; .NET CLR 3.5.30729; .NET CLR 3.0.30729; Media Center PC 6.0; .NET4.0C; .NET4.0E; QQBrowser/7.0.3698.400)",
    "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1; SV1; QQDownload 732; .NET4.0C; .NET4.0E)",
    "Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 5.1; Trident/4.0; SV1; QQDownload 732; .NET4.0C; .NET4.0E; 360SE)",
    "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1; SV1; QQDownload 732; .NET4.0C; .NET4.0E)",
    "Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 6.1; WOW64; Trident/5.0; SLCC2; .NET CLR 2.0.50727; .NET CLR 3.5.30729; .NET CLR 3.0.30729; Media Center PC 6.0; .NET4.0C; .NET4.0E)",
    "Mozilla/5.0 (Windows NT 5.1) AppleWebKit/537.1 (KHTML, like Gecko) Chrome/21.0.1180.89 Safari/537.1",
    "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.1 (KHTML, like Gecko) Chrome/21.0.1180.89 Safari/537.1",
    "Mozilla/5.0 (iPad; U; CPU OS 4_2_1 like Mac OS X; zh-cn) AppleWebKit/533.17.9 (KHTML, like Gecko) Version/5.0.2 Mobile/8C148 Safari/6533.18.5",
    "Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:2.0b13pre) Gecko/20110307 Firefox/4.0b13pre",
    "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:16.0) Gecko/20100101 Firefox/16.0",
    "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.11 (KHTML, like Gecko) Chrome/23.0.1271.64 Safari/537.11",
    "Mozilla/5.0 (X11; U; Linux x86_64; zh-CN; rv:1.9.2.10) Gecko/20100922 Ubuntu/10.10 (maverick) Firefox/3.6.10",
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.36",
]


def get_html_content(original_url):
    ua = random.choice(user_agent_list)
    headers = {'User-Agent': ua}
    request = urllib.request.Request(url=original_url, headers=headers)
    response = urllib.request.urlopen(request, timeout=1000).read()
    #  print("response",response)
    code_info = chardet.detect(response)
    code = code_info['encoding']
    # print("codecodecodeXXXXXXXXXX",code)
    if code == 'utf-8' or code == 'utf-8':
        data = response
    elif code == 'gbk' or code == 'GBK':
        data = response.decode('gbk', 'ignore').encode('utf-8')
    elif code == 'gb2312' or code == 'GB2312':
        data = response.decode('gb2312', 'ignore').encode('utf-8')
    else:
        data = response
    return data


def getLatestNoLast(forum_id):
    original_url = Community_getLatestPost_Url+'&forum_id=' + str(
        forum_id) + '&order=publish_date_desc&size=20'

    data = get_html_content(original_url).decode("utf-8")
    Json = json.loads(data)

    # print("Json", Json)

    data = Json.get("data")
    topics = data.get("topics")

    return topics


def getLatest(forum_id, lastPost):
    original_url = Community_getLatestPost_Url+'&forum_id=' + str(
        forum_id) + '&order=publish_date_desc&size=20&last=' + str(lastPost)

    data = get_html_content(original_url).decode("utf-8")
    Json = json.loads(data)

    # print("Json", Json)

    data = Json.get("data")
    topics = data.get("topics")

    return topics


def GetComment(content):
    # for LSTM
    q = re.findall(r'.{1}', content)
    cut_q = ""
    for word in q:
        cut_q += word + " "

    str = interface(cut_q).replace(" ", "")
    return str





def toComment(topic_id, content, ua, authorization):
    comment = GetComment(content)


    print("content:" + content)
    print("comment:" + comment)

    try:
        comment = input("Do you agree to Comment? Input nothing to confirm or input an appropriate sentence:")
    except InputTimeoutError:
        print('\n Agree to comment')

    print("chatbot will comment on this sentence:"+comment)
    signal.alarm(0)  




    url = Community_toComment_Url
    postdata = {"content": comment, "referenced_id": -1,
                "diff_data": 1,
                "is_watermark": True,
                "topic_id": topic_id}

    para = json.dumps(postdata)

    req = urllib.request.Request(url, para.encode(encoding='UTF8'))
    req.add_header("User-Agent", ua)
    req.add_header("Authorization", authorization)
    req.add_header("Content-Type", "application/json;charset=UTF-8")
    data = urllib.request.urlopen(req).read().decode("utf-8")

    print(data)

    return comment


def save():
    return


def tid_maker():
    return '{0:%Y%m%d%H%M%S%f}'.format(datetime.datetime.now()) + ''.join(
        [str(random.randint(1, 10)) for i in range(5)])





def getPostDetailNoLast(topicId, order_by):
    original_url = Community_getPostDetail_Url + '&topic_id=' + str(
        topicId)
    # print("getPostDetail:"+original_url)
    data = get_html_content(original_url).decode("utf-8")
    Json = json.loads(data)
    data = Json.get("data")
    topic = data.get("topic")
    reviews = data.get("reviews")

    return topic, reviews




cnn_model = CnnModel()



def main():

    nowTime = datetime.datetime.now()


    currentObeserveTime = OBSERVE_INTERVAL + 1  # 

    forumList = [15, 32, 33, 70, 47]
    lastobservedTime15 = nowTime
    lastobservedTime32 = nowTime
    lastobservedTime33 = nowTime
    lastobservedTime70 = nowTime
    lastobservedTime47 = nowTime

    study_startTime = nowTime


    filename = tid_maker()
    isComm = True  #

    obeserveList = []

    time.sleep(1)

    count = 0

    with open('../study/reply/' + filename + "_5" + '.txt', 'w') as f_reply:
        with open('../study/noreply/' + filename + "_5" + '.txt', 'w') as f_noreply:
            with open('../study/hasOthers/' + filename + "_5" + '.txt', 'w') as f_hasOthers:
                while True:

                    ua = random.choice(user_agent_list)
                    ran = random.choice(AI_auth_list)
                    AIName = ran[0]
                    authorization = ran[1]
                    # print(AIName +" "+ authorization+" "+ua)

                    if int((datetime.datetime.now() - study_startTime).seconds / 60) > STUDY_TIME:
                        print("Deployment End ")
                        print(int((datetime.datetime.now() - study_startTime).seconds / 60))
                        for postId in obeserveList:
                            print(postId)
                        break

                    try:

                        if currentObeserveTime > OBSERVE_INTERVAL:
                            print("Time to get latest post")

                            # refresh new posts 刷一遍新出的帖子
                            for forumId in forumList:
                                print("forumId:", forumId)

                                if forumId == 15:
                                    lastobservedTime = lastobservedTime15
                                elif forumId == 32:
                                    lastobservedTime = lastobservedTime32
                                elif forumId == 33:
                                    lastobservedTime = lastobservedTime33
                                elif forumId == 70:
                                    lastobservedTime = lastobservedTime70
                                elif forumId == 47:
                                    lastobservedTime = lastobservedTime47

                                nowobservedTime = datetime.datetime.now()
                                NewestObservedTime = nowobservedTime
                                lastPostId = ""

                                while nowobservedTime > lastobservedTime:
                                    # get new posts 
                                    if lastPostId == "":
                                        topics = getLatestNoLast(forumId)
                                    else:
                                        topics = getLatest(forumId, lastPostId)
                                    for post in topics:

                                        # filter old posts 
                                        published_date = post.get("published_date")

                                        nowobservedTime = datetime.datetime.strptime(published_date,
                                                                                       '%Y-%m-%d %H:%M:%S')
                                        if nowobservedTime < lastobservedTime:
                                            print("has observed!")

                                            break

                                        postId = post.get("id")
                                        lastPostId = postId

                                        # detect if anyone else has replied to this post 看看有没有人回复,有则写入hasOthers
                                        total_review = post.get("total_review")

                                        if total_review == 0:
                                            print("none-replied post")

                                            # 进入观察队列
                                            obeserveList.append(postId)

                                        else:
                                            print("others has replied")

                                            # 已经有人回复了
                                            user_id = post.get("user_id")
                                            title = post.get("title")
                                            forum_id = post.get("forum_id")
                                            f_hasOthers.write(
                                                str(postId) + "\t" + str(forum_id) + "\t" + title.strip().replace("\n",
                                                                                                                  "").replace(
                                                    "\r",
                                                    "") + "\t" + str(user_id) + "\t" + str(
                                                    total_review) + "\t" + published_date + "\n")

                                    if nowobservedTime < lastobservedTime:
                                        break
                                if forumId == 15:
                                    lastobservedTime15 = NewestObservedTime
                                elif forumId == 32:
                                    lastobservedTime32 = NewestObservedTime
                                elif forumId == 33:
                                    lastobservedTime33 = NewestObservedTime
                                elif forumId == 70:
                                    lastobservedTime70 = NewestObservedTime
                                elif forumId == 47:
                                    lastobservedTime47 = NewestObservedTime

                        if currentObeserveTime >= COMMENT_INTERVAL:
                            print("time to check observed posts")

                            print("current observed posts：")
                            # 检查观察队列是否要超过THRESHOLD的
                            for postId in obeserveList[0:]:
                                print(postId)

                                try:
                                    topic, reviewers = getPostDetailNoLast(postId, "reviewed_date")
                                except Exception as e:
                                    print(e)
                                    print(" There is an exception when getting post detail,so remove this post in observed posts list")
                                    obeserveList.remove(postId)
                                    print("remove ", postId)
                                    continue

                                time.sleep(2)
                                total_floor = topic.get("total_floor")
                                published_date = topic.get("published_date")
                                user_id = topic.get("user_id")
                                title = topic.get("title")
                                if title is None:
                                    title = "No Title"
                                forum_id = topic.get("forum_id")
                                total_review = topic.get("total_review")
                                content = topic.get("content")
                                user_name = topic.get("publisher").get("screen_name")
                                if total_floor > 0:
                                    print("others has replied")
                                    f_hasOthers.write(
                                        str(postId) + "\t" + str(forum_id) + "\t" + title.strip().replace("\n",
                                                                                                          "").replace(
                                            "\r",
                                            "") + "\t" + str(user_id) + "\t" + str(
                                            total_review) + "\t" + published_date + "\n")
                                    obeserveList.remove(postId)
                                    print("remove ", postId)
                                    continue
                                timeDiff = int((datetime.datetime.now() - datetime.datetime.strptime(published_date,
                                                                                                     '%Y-%m-%d %H:%M:%S')).seconds / 60)
                                print("timeDiff:", timeDiff)
                                if timeDiff > THRESHOLD:
                                    cat = cnn_model.predict(content)
                                    print("cat", cat)
                                    if cat == "0":
                                        print("cat:0")
                                        f_noreply.write(
                                            "reason:info" + "\t" + str(postId) + "\t" + str(
                                                forum_id) + "\t" + title.strip().replace(
                                                "\n", "").replace("\r", "") + "\t" + content.strip().replace("\n",
                                                                                                             "").replace(
                                                "\r", "") + "\t" +
                                            str(user_id) + "\t" + str(total_review) + "\t" + published_date + "\n")
                                        obeserveList.remove(postId)
                                        print("remove ", postId)
                                        continue

                                    print("cat:1")
                                    # Randomly select half of the posts to reply 达到回复要求，选择half去回复
                                    try:
                                        if isComm:
                                            print("count", count)
                                            count += 1

                                            comment = toComment(postId, content, ua, authorization)
                                            print("will to comment:", AIName + " " + comment)
                                            if comment != "":
                                                logstr = "postId:" + str(
                                                    postId) + "\tAIName:" + AIName + "\tforum_id:" + str(
                                                    forum_id) + "\tuser_id:" + str(user_id) + \
                                                         "\tuser_name:" + user_name + "\ttitle:" + title.strip().replace(
                                                    "\n", "").replace("\r",
                                                                      "") + "\tcontent:" + content.strip().replace("\n",
                                                                                                                   "").replace(
                                                    "\r", "") + "\tdata：" + published_date + \
                                                         "\ttotal_floor:" + str(
                                                    total_floor) + "\tcomment:" + comment + "\ttimeDiff:" + str(
                                                    timeDiff) + "\n"

                                                f_reply.write(logstr)
                                        else:
                                            print("randomly not to reply")
                                            logstr = "reason:randon" + "\tpostId:" + str(postId) + "\tforum_id:" + str(
                                                forum_id) + "\tuser_id:" + str(user_id) + \
                                                     "\tuser_name:" + user_name + "\ttitle:" + title.strip().replace(
                                                "\n", "").replace("\r", "") + "\tcontent:" + content.strip().replace(
                                                "\n", "").replace("\r", "") + "\tdata：" + published_date + \
                                                     "\ttotal_floor:" + str(total_floor) + "\n"
                                            f_noreply.write(logstr)
                                        isComm = not isComm
                                    except Exception as e:
                                        print("Exception", e)
                                        print("AIName", AIName)

                                    obeserveList.remove(postId)
                                    print("remove ", postId)
                            print("current observed posts list：")
                            for postId in obeserveList:
                                print(postId)

                        if currentObeserveTime > OBSERVE_INTERVAL:
                            currentObeserveTime = 0

                        print("sleep with Comment_interval")
                        time.sleep(COMMENT_INTERVAL * 60)
                        currentObeserveTime += COMMENT_INTERVAL
                        print("currentObserveTime", currentObeserveTime)



                    except Exception as e:
                        print('Error:', e)
                        print("wrong！")
                        print("file save as", filename)
                        currentObeserveTime += COMMENT_INTERVAL
                        print("currentObserveTime", currentObeserveTime)

            f_hasOthers.close()
        f_noreply.close()
    f_reply.close()

    print("file save as", filename)
    print("end")


if __name__ == '__main__':
    main()
